### ISSUE TITLE
[Not an issue: a Question] Client to connect to both ES 5 and ES7.

### ISSUE THREAD

**Issue by parvathi-nair** [olivere/elastic#1536]
### ES Versions
elastic.v7 (for Elasticsearch 7.x) AND
elastic.v5 (for Elasticsearch 5.x)

### Please describe the expected behavior
We have a project that is currently using ES 5.5. We are in the process of migrating to ES 7.10. However until all the migrations are completed, we are planning to have both instances of the ES present in our environment.

Our Go code would need to connect to both ES 5 and ES7 based on the server it is talking to. 

We had glide dependency for ES 5.5 version client. We are trying to add ES 7 version of client for ES7. Is that supported?

Is there a way I can use both client versions? It says only go mod is supported for 7. But I was just wondering whether there is a way.

- package: gopkg.in/olivere/elastic.v5
  version: =v5.0.60

Trial 1: Tried with this in glide.yaml
- package: github.com/olivere/elastic/v7
   version: 593cd3205b5bf747034445b8230e20b29edf57e4

Trial 2: Tried with this in glide.yaml
- package: gopkg.in/olivere/elastic.v7
  version: v7.0.29
  repo: https://github.com/olivere/elastic

We get error in both cases with v7 path. since the glide cache does not  have v7 in its path. It is directly underneath .glide/cache/src/https-github.com-olivere-elastic

Error scanning github.com/olivere/elastic/v7/config: open src/.glide/cache/src/https-github.com-olivere-elastic/v7/config: no such file or directory.

Has anyone come across this same use-case? Any ideas on how to resolve this?

**Comment by olivere** [comment]
Hello! 

tldr: If you keep yourself using only the stable and unchanged APIs of ES, you may be able to use v7 of elastic with e.g. v5 of elastic. However, there might be dragons, and some things might unexpectedly _not_ work and break at runtime.

The slightly longer story:

What my team and I found valuable after working on some of the same problems that you describe:

1. **Switch to [Go modules](https://go.dev/blog/using-go-modules).** Remove `dep` and `glide`. Yes, some people don't like it. But it's the way Go, the language, has adopted for managing dependencies. If you don't use it, you keep fighting the language. It's a fight that's not worth it.
2. **Encapsulate as much as you can.** Go interfaces are great. Create the implementation at one place in your code (e.g. in `main`, then keep yourself to only pass the interface around. In other words: [Wire your components explicitly](https://peter.bourgon.org/go-for-industrial-programming/#the-component-graph). If you need to have two implementations of an interface, feel free to use [a kind of Factory pattern](https://www.sohamkamani.com/golang/2018-06-20-golang-factory-patterns/).

Here's an example of what I meant by hiding the implementation behind an interface. In the [`esdiff` repository](https://github.com/olivere/esdiff), I needed to write code that iterates on an Elasticsearch indices for different versions of ES. So I [abstracted the code to iterate into an interface](https://github.com/olivere/esdiff/blob/master/elastic/client.go#L9-L12) (nowadays I would've probably encapsulated even more and build different interfaces, e.g. `Iterator`, `Indexer` etc.). But anyway, here's the interface:

```go
// Client encapsulates access to an Elasticsearch cluster.
type Client interface {
	Iterate(context.Context, *IterateRequest) (<-chan *diff.Document, <-chan error)
}
```

Now I have 3 implementations of that interfaces, one for each ES version. You can find them in the directories [`v5`](https://github.com/olivere/esdiff/tree/master/elastic/v5), [`v6`](https://github.com/olivere/esdiff/tree/master/elastic/v6), and [`v7`](https://github.com/olivere/esdiff/tree/master/elastic/v7). The logic to create a new client based on the Elasticsearch version is [done once and only once, in `main`](https://github.com/olivere/esdiff/blob/master/main.go#L137-L179). Of course, you might also do that at runtime, but the more important thing to notice is that _no other code outside [this method](https://github.com/olivere/esdiff/blob/master/main.go#L137-L179)_ will __ever__ use the directories `v5`, `v6`, or `v7`: The only thing they know is the interface `Client`.

By encapsulating the implementation and hiding it behind an interface, you can pass around `Client` interface without the caller even knowing that there is an ES behind it (and which version). You could even switch it to some different storage option, given it implements the interface.
 
HTH

**Comment by parvathi-nair** [comment]
Thanks for your reply. We are not able to switch to GO modules yet due to n number of other go dependencies in our service.  Will try with V6 client and hoping that would be good enough to talk to Elasticsearch 7 and Elasticsearch 5.5.
